pkg/declarativeconfig/context.go

```diff
@@ -0,0 +1,26 @@
+package declarativeconfig
+
+import (
+	"context"
+
+	"github.com/stackrox/rox/generated/storage"
+)
+
+type originCheckerKey struct{}
+
+// AllowOnlyDeclarativeOperations signals that the context holder is allowed to modify declarative resources.
+const AllowOnlyDeclarativeOperations = true
+
+// WithAllowOnlyDeclarativeOperations returns a context that is a child of the given context and allows to modify
+// declarative resources.
+func WithAllowOnlyDeclarativeOperations(ctx context.Context) context.Context {
+	return context.WithValue(ctx, originCheckerKey{}, AllowOnlyDeclarativeOperations)
+}
+
+// IsOriginModifiable returns whether context allows to modify declarative resources.
+func IsOriginModifiable(ctx context.Context, origin storage.Traits_Origin) bool {
```

Comment:
Maybe it would be worthwhile to do the following changes:

Instead of passing in `Traits_Origin`, since clients won't actually care much about the `Traits` but rather the resource, we can create an interface and let clients pass in their proto message, and we'll receive the traits origin within the function:
```go
type ResourceWithTraits interface {
  GetTraits() *storage.Traits
}

func CanModifyResource(ctx context.Context, resource ResourceWithTraits) bool {
	if ctx.Value(originCheckerKey{}) == AllowOnlyDeclarativeOperations {
		return resource.GetTraits().GetOrigin() == storage.Traits_DECLARATIVE
	}
	return resource.GetTraits().GetOrigin() == storage.Traits_IMPERATIVE
}

// Clients can then call it as such:
func verifyPermissionSetOriginMatches(ctx context.Context, ps *storage.PermissionSet) error {
	if !declarativeconfig.CanModifyResource(ctx, ps) {
		return errors.Wrapf(errox.InvalidArgs, "permission set %q's origin is %s, cannot be modified or deleted within this context",
			ps.GetName(), ps.GetTraits().GetOrigin())
	}
	return nil
}
```

I think that ones more worthwhile to have - it's also _nearly_ in line with all our protos we have in scope for declarative config, as they do not define the `Traits` message nested. Exception would be the `Groups` proto, but I'm willing to make the compromise since its "weird" per se - defining the ID within the `GroupProperties` message.